pageserver/src/layered_repository/snapshot_layer.rs

```diff
@@ -272,21 +326,31 @@ impl Layer for SnapshotLayer {
     ) -> Result<Option<Lsn>> {
         // Scan the BTreeMap backwards, starting from the given entry.
         let mut need_base_image_lsn: Option<Lsn> = Some(lsn);
+
+        // TODO: avoid opening the snapshot file for each read
```

Comment:
I previously tried caching the book in `SnapshotLayerInner` but tests were failing locally due to running out of file descriptors.

Reply:
Ah yeah. Postgres works around that problem with so-called "virtual file descriptors". They are used instead of real file descriptors, with wrapper functions around the usual read/write calls. Whenever you call read on a virtual file descriptor, it checks if the real file is open and opens it if it's not. And if too many files are open, it closes the underlying real file descriptor of some other virtual file descriptor, with a simple Least-Recently-Used Cache to decide which underlying files to keep open. We may need something similar. (as a separate task and PR)